pkgs/applications/misc/dmenu-rs/default.nix

```diff
@@ -0,0 +1,111 @@
+{ lib
+, fetchFromGitHub
+, rustPlatform
+, clang
+, libclang
+, llvm
+, m4
+, pkg-config
+, expat
+, fontconfig
+, libX11
+, libXft
+, libXinerama
+, python3
+}:
+
+# The dmenu-rs package has extensive plugin support. However, this derivation
+# only builds the package with the default set of plugins. If you'd like to
+# further customize dmenu-rs you can build it from the source.
+# See: https://github.com/Shizcow/dmenu-rs#plugins
+rustPlatform.buildRustPackage rec {
+  pname = "dmenu-rs";
+  version = "5.5.1";
+
+  src = fetchFromGitHub {
+    owner = "Shizcow";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-WpDqBjIZ5ESnoRtWZmvm+gNTLKqxL4IibRVCj0yRIFM=";
+  };
+
+  nativeBuildInputs = [
+    clang
+    libclang
+    llvm
+    m4
+    pkg-config
+    python3
+  ];
+
+  buildInputs = [
+    expat
+    fontconfig
+    libX11
+    libXft
+    libXinerama
+  ];
+
+  # The dmenu-rs repository does not include a Cargo.lock because of its
+  # dynamic build and plugin support. Generating it with make and checking it
+  # in to nixpkgs here was the easiest way to supply it to buildRustPackage.
+  # See: https://github.com/Shizcow/dmenu-rs/issues/34#issuecomment-757415584
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+  };
+
+  # Several nixpkgs' functions from its rust build tooling expect cargo files
+  # to exist at the project's root directory unless otherwise specified.
+  cargoRoot = "src";
+
+  # The rust-xcb dependency dynamically generates rust code at build time.
+  # This derivation uses nixpkgs rust functions that vendor each cargo
+  # dependency's source code into the READ-ONLY nix store. To avoid the code
+  # generation step failing, we copy the rust-xcb source out of the nix store
+  # and make it writeable.
+  # See: https://github.com/rust-x-bindings/rust-xcb/tree/v0.8.2
+  patchPhase = ''
+    chmod +w /build/cargo-vendor-dir
+    mkdir -p /build/cargo-vendor-dir/xcb-0.8.2-readwrite
+    cp --recursive /build/cargo-vendor-dir/xcb-0.8.2/. /build/cargo-vendor-dir/xcb-0.8.2-readwrite
+    unlink /build/cargo-vendor-dir/xcb-0.8.2
+    mv /build/cargo-vendor-dir/xcb-0.8.2-readwrite /build/cargo-vendor-dir/xcb-0.8.2
+    chmod -R +w /build/cargo-vendor-dir/xcb-0.8.2/
+  '';
+
+  # As suggested in the nixpkgs manual section on building rust packages, the
+  # derivation's Cargo.lock file is copied into the source.
+  postPatch = ''
+    cp ${./Cargo.lock} src/Cargo.lock
+  '';
+
+  preConfigure = ''
+    sed -i "s@PREFIX = /usr/local@PREFIX = $out@g" config.mk
+  '';
+
+  buildPhase = ''
+    make
+  '';
+
+  installPhase = ''
+    make install
+  '';
+
+  # Tests require an X11 server and yields a cannot open display error. It's
+  # probably possible to get an X server running in the build but seems hard.
+  doCheck = false;
+  checkPhase = ''
+    make test
+  '';
```

Comment:
While I'd like to enable this, the test triggered by `make test` invokes `dmenu` directly.

[It runs][1]

```bash
seq 1 100 | target/dmenu $(ARGS)
```

This not only requires an X11 server but it also hangs waiting for user input.

I was able to hack together the following `checkPhase` that works locally using `timeout`

```bash
timeout --preserve-status 5s xvfb-run make test
```

however simply hoping that 5 seconds is long enough for xvfb-run to spin up and the test to complete on every machine seems like an awful idea. I'm not sure how to kill the hanging call to `dmenu`, no option seems to allow that and I don't know how to feed in a keypress like `<RETURN>` into the `xvfb-run` context.

So I'm just leaving it disabled and relying on the other tests like installing, running the binary, and using it personally on a regular basis.